### PR TITLE
Allow directly passing text embeddings to Stable Diffusion Pipeline for prompt weighting

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR allows to pass text_embeddings directly to the SD pipelines which should greatly help with writing pipelines such as Weight Prompting, Prompt-2-Prompt, etc...

For more details see discussion here: https://github.com/huggingface/diffusers/issues/1506#issuecomment-1399169567

As it's always the same function (`self._encode_prompt`), two tests for stable diffusion should be enough here IMO.

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py [hunk 1]
```diff
@@ -302,12 +334,15 @@ def prepare_extra_step_kwargs(self, generator, eta):
...
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion_img2img.StableDiffusionImg2ImgPipeline.check_inputs
+    def check_inputs(
+        self, prompt, strength, callback_steps, negative_prompt=None, prompt_embeds=None, negative_prompt_embeds=None
+    ):
         if not isinstance(prompt, str) and not isinstance(prompt, list):
             raise ValueError(f"`prompt` has to be of type `str` or `list` but is {type(prompt)}")
 
         if strength < 0 or strength > 1:
-            raise ValueError(f"The value of strength should in [1.0, 1.0] but is {strength}")
+            raise ValueError(f"The value of strength should in [0.0, 1.0] but is {strength}")
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_depth2img.py:345]: 👍

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py [hunk 2]
```diff
@@ -491,15 +543,19 @@ def _encode_prompt(self, prompt, device, num_images_per_prompt, do_classifier_fr
...
+            uncond_embeddings = uncond_embeddings.to(dtype=self.text_encoder.dtype, device=device)
+
             uncond_embeddings = uncond_embeddings.repeat(1, num_images_per_prompt, 1)
             uncond_embeddings = uncond_embeddings.view(batch_size * num_images_per_prompt, seq_len, -1)
 
             # For classifier free guidance, we need to do two forward passes.
             # Here we concatenate the unconditional and text embeddings into a single batch
             # to avoid doing two forward passes
-            text_embeddings = torch.cat([text_embeddings, uncond_embeddings, uncond_embeddings])
+            text_embeddings = torch.cat([uncond_embeddings, text_embeddings])
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py:558]: I think this is a typo? InstructPix2Pix requires 3 sets of latents, and the order is different in this pipeline for some reason (it's text + uncond + uncond whereas in the others it's uncond + text).

**Review by pcuenca — COMMENTED** [review]
I reviewed the first pipeline in depth, the comments would apply to others. I believe there's a typo in the InstructPix2PixPipeline?

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py [hunk 3]
```diff
@@ -236,44 +245,63 @@ def _encode_prompt(self, prompt, device, num_images_per_prompt, do_classifier_fr
             negative_prompt (`str` or `List[str]`):
                 The prompt or prompts not to guide the image generation. Ignored when not using guidance (i.e., ignored
                 if `guidance_scale` is less than `1`).
+            text_embeddings (`torch.FloatTensor`, *optional*):
+                Pre-generated text embeddings. Can be used to easily tweak text inputs, *e.g.* prompt weighting. If not
+                provided, text embeddings will be generated from `prompt` input argument.
+            uncond_embeddings (`torch.FloatTensor`, *optional*):
+                Pre-generated negative text embeddings. Can be used to easily tweak text inputs, *e.g.* prompt
+                weighting. If not provided, negative_prompt_embeds will be generated from `negative_prompt` input
+                argument.
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:254]: +1 for the first point.

For me, `text_embeddings` is fine or another option would be `prompt_embeddings`

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py [hunk 4]
```diff
@@ -236,44 +245,63 @@ def _encode_prompt(self, prompt, device, num_images_per_prompt, do_classifier_fr
...
-            return_tensors="pt",
-        )
-        text_input_ids = text_inputs.input_ids
-        untruncated_ids = self.tokenizer(prompt, padding="longest", return_tensors="pt").input_ids
+        if prompt is not None and isinstance(prompt, str):
+            batch_size = 1
+        elif prompt is not None and isinstance(prompt, list):
+            batch_size = len(prompt)
+        else:
+            batch_size = text_embeddings.shape[0]
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:261]: +1, I think we can simply check if both are `None` and raise if `True`

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py [hunk 5]
```diff
@@ -313,8 +341,12 @@ def _encode_prompt(self, prompt, device, num_images_per_prompt, do_classifier_fr
             )
             uncond_embeddings = uncond_embeddings[0]
 
+        if do_classifier_free_guidance:
             # duplicate unconditional embeddings for each generation per prompt, using mps friendly method
             seq_len = uncond_embeddings.shape[1]
+
```
[the remark this answers is not in the document]
Comment (patil-suraj) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:347]: +1

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py [hunk 6]
```diff
@@ -375,6 +413,24 @@ def check_inputs(self, prompt, height, width, callback_steps):
                 f" {type(callback_steps)}."
             )
 
+        if prompt is not None and prompt_embeds is not None:
+            raise ValueError(
+                f"Cannot forward both `prompt`: {prompt} and `prompt_embeds`: {prompt_embeds}. Please make sure to"
+                " only forward one of the two."
```
Comment (patil-suraj) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:419]: maybe 
```suggestion
                "You can only specify one of `prompt` and `prompt_embeds`. Please make sure to only specify one of them."
```

Think it's better to have shorter errors.

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py [hunk 7]
```diff
@@ -375,6 +413,24 @@ def check_inputs(self, prompt, height, width, callback_steps):
...
+            raise ValueError(
+                "Provide either `prompt` or `prompt_embeds`. Cannot leave both `prompt` and `prompt_embeds` undefined."
+            )
+        elif prompt is not None and (not isinstance(prompt, str) and not isinstance(prompt, list)):
+            raise ValueError(f"`prompt` has to be of type `str` or `list` but is {type(prompt)}")
+
+        if negative_prompt is not None and negative_prompt_embeds is not None:
+            raise ValueError(
+                f"Cannot forward both `negative_prompt`: {negative_prompt} and `negative_prompt_embeds`:"
+                f" {negative_prompt_embeds}. Please make sure to only forward one of the two."
```
Comment (patil-suraj) [inline src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:431]:
```suggestion
                "You can only specify one of `negative_prompt` and `negative_prompt_embeds`. Please make sure to only specify one of them."
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py (hunk 2 above)
Comment (patil-suraj) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py:558]: Good catch! And yes, the order follows the original implem.

**Review by patil-suraj — COMMENTED** [review]
Looks good. I just left some nits. And as Pedro said, for `InstrcutPix2Pix`, we need uncond embedding concat is different from other pipelines.

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:254]: Yeah that's a good point! I'll change the naming to `prompt_embeddings` everywhere

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:261]: This is actually done in `check_inputs` now if I'm not mistaken :-)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py:558]: It's because I added the Copy-from statement, I order them correctly after the function :-)

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:419]: Disagree here, people will recognize easily what they have passed, so think it's always good to be specific

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pcuenca) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:261]: 👍

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py [hunk 8]
```diff
@@ -243,7 +255,18 @@ def __call__(
...
-            prompt, device, num_images_per_prompt, do_classifier_free_guidance, negative_prompt
+            prompt,
+            device,
+            num_images_per_prompt,
+            do_classifier_free_guidance,
+            negative_prompt,
+            text_embeddings=prompt_embeds,
+            uncond_embeddings=negative_prompt_embeds,
+        )
+        # pix2pix has two uncond_embeddings
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py:266]:
```suggestion
        # pix2pix has two uncond_embeddings, and unlike in other pipelines latents are ordered [text_embeddings, uncond_embeddings, uncond_embeddings]
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py (hunk 2 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py:558]: Oh, I see. Yeah, I missed that. I'm hesitant about the convenience of adding the copied-from comment vs the potential loss in legibility. I suggested a more verbose comment above.

**Review by pcuenca — APPROVED** [review]

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py:558]: Yeah it's not ideal, I'll get the "# End Copy statement working, so then it should be much more readable

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py:347]: Done.

File: src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/pipelines/alt_diffusion/pipeline_alt_diffusion.py]: Done.
